app/jobs/autoscaling.py

```diff
@@ -260,227 +61,313 @@ def get_instance_tags():
-    if instance_ids:
-        print "Starting to drain the following instances: " + ",".join(instance_ids)
-        add_draining_tag(instance_ids, draining_tag)
+def autoscaling_update(my_num_jobs, my_environment="development",
+                       max_job_dispatch_lag_seconds=900, job_tag_prefix="RunningIDseqBatchJob_",
+                       job_tag_keep_alive_seconds=600, draining_tag="draining"):
+    '''
+    Plan and perform an update to the GSNAP/RAPSEARCH ASGs.
+    Note: Will not scale up for jobs originating in a development environment.
+          You can provision machines for development jobs by manually setting MinSize on the ASGs,
+          which is respected by this autoscaler.
+    '''
 
+    asg_list = get_asg_list()
+    tag_list = get_tag_list()
+    gsnap_ASG = ASG("gsnapl", my_environment, asg_list, tag_list, draining_tag, job_tag_prefix, max_job_dispatch_lag_seconds, job_tag_keep_alive_seconds)
+    rapsearch_ASG = ASG("rapsearch2", my_environment, asg_list, tag_list, draining_tag, job_tag_prefix, max_job_dispatch_lag_seconds, job_tag_keep_alive_seconds)
+    if not (gsnap_ASG.can_scale and rapsearch_ASG.can_scale):
+        print "Scaling by agents of {my_environment} is not permitted.".format(my_environment=my_environment)
+        return
 
-def stop_draining(asg, tag_list, draining_tag, num_instances):
-    instance_ids = instances_to_rescue(asg, tag_list, draining_tag, num_instances)
-    if instance_ids:
-        print "Stopping drainage of the following instances: " + ",".join(instance_ids)
-        remove_draining_tag(instance_ids, draining_tag)
+    print "{my_num_jobs} jobs are in progress.".format(my_num_jobs=my_num_jobs)
+    compute_num_instances = required_capacity(my_num_jobs)
+    for service_ASG in [gsnap_ASG, rapsearch_ASG]:
+        print "--- Analyzing the {instance_name} ASG ---".format(instance_name=service_ASG.instance_name)
+        num_desired = service_ASG.num_desired_instances()
+        healthy_instances, draining_instances, terminating_instances, corrupt_instances = service_ASG.classify_instances()
+        num_healthy = len(healthy_instances)
+        num_draining = len(draining_instances)
+        print "CURRENTLY:"
+        print "The desired capacity is {num_desired}.".format(num_desired=num_desired)
+        print "There are {num_healthy} healthy instances: {healthy_instances}.".format(num_healthy=num_healthy, healthy_instances=healthy_instances)
+        print "There are {num_draining} draining instances: {draining_instances}.".format(num_draining=num_draining, draining_instances=draining_instances)
+        print "There are {num_terminating} terminating instances: {terminating_instances}.".format(num_terminating=len(terminating_instances), terminating_instances=terminating_instances)
+        assert not corrupt_instances, "There are {num_corrupt} corrupt instances: {corrupt_instances}.".format(num_corrupt=len(corrupt_instances), corrupt_instances=corrupt_instances)
```

Comment:
Are you sure you want to crash here?   Who would notice this?

Reply:
good point. I will just print a warning. I have not encountered this case in my testing and doubt it exists. In any case, do not want to crash.